src/ILLink.Shared/Dataflow/ITransfer.cs

```diff
@@ -0,0 +1,32 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace ILLink.Shared
+{
+	// ITransfer represents the transfer functions for a dataflow analysis.
+	// The transfer functions compute the effects of an operation on the set of facts
+	// tracked by a dataflow analysis. This simulates the execution of the operation
+	// on the domain of abstract values tracked in the dataflow analysis.
+
+	// TValue is the type of the information tracked in a dataflow analysis at each program point.
+	// TLattice is the type of the lattice formed by these values.
+
+	// TOperation is a type representing the operations that the transfer function
+	// "simulates". It isn't constrained by the interface, but is typically a basic block,
+	// where the transfer functions are defined in terms of transfer functions for individual
+	// operations in the block.
+
+	// TLattice isn't typically used in the implementation except to provide the "Top" value.
+	// This expresses the conceptual constraint that the transferred values are part of a lattice.
+	public interface ITransfer<TOperation, TValue, TLattice>
+		where TValue : IEquatable<TValue>
+		where TLattice : ILattice<TValue>
+	{
```

Comment:
I don't think this abstraction is right -- a lattice isn't used by the transfer function, instead the mathematical structure is that

1. A data-flow analysis is a program for calculating the set of possible values
2. It requires a domain, which consists of a set of values being tracked, a control flow graph made of basic blocks, a transfer function between those blocks, a defined top state, and a joining function for control flow edges.

I think this means that the right abstraction combines all these things into one. They wouldn't be seperable:

```C#
interface IDataFlowAnalysis<TValueSet>
{
    public Run(ControlFlowGraph graph)
    {
        ...
    }

    public TValue Transfer (TValue);
    public TValue Top { get; }
    public TValue Meet (TValue, TValue);
}
```

In other words, a lattice is a structure that a data-flow analysis _is_, not something that other things _have_

Reply:
Ah, I see -- you're abstracting away local variable analysis as well, so you need to need two lattice implementations: one for the variable state, and one for the global state.

I'll see if I can come up with a shorter way to express this.

Reply:
Hm, I think the biggest simplification would be getting rid of support for general data flow and just doing DAM, but I'm not sure if there was another data flow analysis we need to support.

The other simplification could be creating an `IDataFlowAnalysis : ILattice` and adding `Transfer` to `IDataFlowAnalysis`. Having those split out doesn't seem to providing a lot of generality.

Reply:
Yes, I'd like to keep the lattice abstraction separate because it allows defining the dictionary lattices in terms of another lattices - I personally think it makes the algebraic structure clearer. One simplification that would be really nice is to combine TValue and TValueLattice, by having an interface like ILatticeValue where Top is a static virtual interface method. But AFAIK we can't do that with netstandard2.0.

I don't think we need to support another analysis yet. If we don't think there's value in abstracting this out I can make everything DAM-specific, but one of my original design goals was to make this a reusable framework (or at least for the core pieces to be a reasonable starting point for such a framework). I also think using these abstractions forces thinking in more mathematical terms which I believe is going to help with correctness - but that might just be the way I think. :)

Reply:
I personally don't mind this abstraction. Precisely for the point that it nicely maps to all of the theoretical work in this space.
I wonder if we could use a similar approach (haven't really thought it through) for constant propagation and branch removal.... in the linker/native AOT.

Reply:
For the second suggestion (combining the flow analysis with the transfer functions under an `IDataFlowAnalysis` or similar): as discussed, we can't use default interface implementations so this would prevent the current approach where the core dataflow analysis is generic. Alternatively, it would require the visitor state to be stored together with the class implementing the dataflow analysis, but this state doesn't need to live as long as the analysis results. For now I've left them separate but it's easy enough to change if we want to choose a different tradeoff later.